### ISSUE TITLE
fused operator splitting issue

### ISSUE THREAD

**Issue by V1ewLine** [microsoft/nnscaler#6]
After reviewing the runtime logs, I observed that the fused operator 'flash attention' is also split. Since 'flash attention' involves CUDA kernels, I wonder if splitting this operator could affect its runtime efficiency？Is it possible that the performance of flash attention becomes worse after splitting ？

**Comment by zyeric** [comment]
Assume the example you run is `llama3_8B_128K`.

In this example, flash attention is registered with annotation `b l^ (group_num 4) hd^, b s^ group_num hd^, b s^ (group_num 4) vd^ -> b l^ (group_num 4) vd^` to nnscaler.

This annotation denotes that flash-attention can be split along batch dim `b` and head group dim `group_num`.

After the computation graph has been built, nnscaler's plan searcher `AutoDist` will profile all the possible partition of the flash-attn. For example, if there are 4 devices, and the input shape of q, k, v is [4, 2048, 32, 128], [4, 2048, 8, 128] and [4, 2048, 8, 128], `AutoDist` will profile the operator with initial shape and with 2 partitioned shapes ([4, 2048, 8, 128], [4, 2048, 2, 128], [4, 2048, 2, 128]) and ([1, 2048, 32, 128], [1, 2048, 8, 128], [1, 2048, 8, 128]). After that, it will determine the partition by solving a global optimization problem which considers profile numbers of each (partitioned) operator and communication cost.

In conclusion, if the input shape is sufficiently large for flash-attn (for example, the sequence length is 128K), the MFU of the initial operator and partitioned one should be close since the GPU has be saturated by the computation. If the input shape is small, the MFU of the partitioned flash-attn may be lower, however

- autodist may not partition the operator based on the profile data
- although the MFU is lower, but the end2end training speed is faster, and autodist will partition the operator as well

**Comment by V1ewLine** [comment]
I see，thank you for your explanation~
